File: docs/chapters/chapter-4/assets/case-THU-DA-L02-C01-A03/r/objects.R
# 同一CSV与同一筛选规则；基础R，不安装包。
script <- sub("^--file=", "", grep("^--file=", commandArgs(), value=TRUE)[1])
base <- dirname(dirname(normalizePath(script, mustWork=TRUE)))
frame <- read.csv(file.path(base, "data/guesses.csv"), colClasses=c("character","character","integer","numeric"), na.strings="", check.names=FALSE)
missing_columns <- function(frame, required) required[!required %in% names(frame)]
stopifnot(length(missing_columns(frame, c("record_id","participant_id","round","estimate_grains"))) == 0)
keep <- !is.na(frame$estimate_grains) & frame$estimate_grains >= 950
selected <- frame[keep, c("record_id","estimate_grains"), drop=FALSE]
metrics <- data.frame(metric=c("records","participants","missing_estimates","selected_records"),
                      value=c(nrow(frame),length(unique(frame$participant_id)),sum(is.na(frame$estimate_grains)),nrow(selected)))
out <- file.path(base, "results/r")
dir.create(out, recursive=TRUE, showWarnings=FALSE)
write.csv(metrics, file.path(out,"metrics.csv"), row.names=FALSE, na="")
write.csv(selected, file.path(out,"selected.csv"), row.names=FALSE, na="")
print(metrics)
cat("selected IDs:", paste(selected$record_id, collapse=","), "\n")

File: docs/chapters/chapter-4/assets/case-THU-DA-L02-C01-A03/python/objects.py
"""同一五行教学表：对象、定位、缺失检查；无统计推断。"""
from pathlib import Path
import pandas as pd

BASE = Path(__file__).resolve().parents[1]

def missing_columns(frame, required):
    return [name for name in required if name not in frame.columns]

def main():
    frame = pd.read_csv(BASE / "data/guesses.csv", dtype={"record_id": str, "participant_id": str})
    assert not missing_columns(frame, ["record_id", "participant_id", "round", "estimate_grains"])
    selected = frame.loc[frame["estimate_grains"].notna() & (frame["estimate_grains"] >= 950), ["record_id", "estimate_grains"]]
    metrics = pd.DataFrame({"metric": ["records", "participants", "missing_estimates", "selected_records"],
                            "value": [len(frame), frame.participant_id.nunique(), int(frame.estimate_grains.isna().sum()), len(selected)]})
    out = BASE / "results/python"
    out.mkdir(parents=True, exist_ok=True)
    metrics.to_csv(out / "metrics.csv", index=False)
    selected.to_csv(out / "selected.csv", index=False)
    print(metrics.to_string(index=False))
    print("selected IDs:", ",".join(selected.record_id))

if __name__ == "__main__":
    main()
